File: Arrays/268_Missing_Number.cpp
/*
https://leetcode.com/problems/missing-number/submissions/
Brute force
TC: O(n)
SC: O(1)
*/

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int n=nums.size();
        for(int i=0;i<n+1;i++)
        {
            if(count(nums.begin(),nums.end(),i)==0)
                return i;
        }
        return -1;
    }
};
/*
Sorting 
*/

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int n=nums.size();
        sort(nums.begin(),nums.end());
        for(int i=0;i<n;i++)
        {
            if(i!=nums[i])
                return i;
        }
        return n;
    }
};
//1.XOR
class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int res = nums.size();
        for(int i=0; i<nums.size(); i++){
            res ^= i;
            res ^= nums[i];
        }
        return res;
    }
};*/

//2.SUM
/*
O(n)
O(1)*/
class Solution {
public:
    int missingNumber(vector<int>& nums) { //sum
        int len = nums.size();
        //sum of all numbers in range len
        int sum = (0+len)*(len+1)/2;
        for(int i=0; i<len; i++)
            sum-=nums[i];
        return sum;
    }
};

/*
//3.Binary Search
TC: O(nlogn)
SC: O(1)
*/
class Solution {
public:
    int missingNumber(vector<int>& nums) {
        //binary search
        sort(nums.begin(),nums.end());
        int left = 0, right = nums.length, mid= (left + right)/2;
        while(left<right){
            mid = (left + right)/2;
            if(nums[mid]>mid) right = mid;
            else left = mid+1;
        }
        return left;
};

File: Blind75/Arrays/268_Missing_Number.py
# #O(3n)
# class Solution:
#     def missingNumber(self, nums: List[int]) -> int:
#         a=max(nums)
#         for i in range(a+2):
#             if(i not in nums):
#                 return i

# class Solution:
#     def missingNumber(self, nums: List[int]) -> int:
#         s=set(nums)
#         n=len(nums)
#         for i in range(n+1):
#             if(i not in s):
#                 return i
      
# #sorting
# class Solution:
#     def missingNumber(self, nums: List[int]) -> int:
#         nums.sort()
#         for i in range(len(nums)):
#             if(i!=nums[i]):
#                 return i
#         return i+1

# class Solution:
#     def missingNumber(self, nums: List[int]) -> int:
#         ans=0
#         n=len(nums)
#         for i in range(1,n+1):
#             ans^=i
#         for i in range(n):
#             ans^=nums[i]
#         return ans


# class Solution:
#     def missingNumber(self, nums: List[int]) -> int:
#         n=len(nums)
#         s1=int(n*(n+1)/2)
#         s2=sum(nums)
#         return s1-s2

class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        ans=len(nums)
        for i in range(len(nums)):
            ans+=(i-nums[i])
        return ans
